### PR TITLE
Added Buffer Copy tests for WebGL 2

### PR DESCRIPTION
**Pull request by marcorperez**
Signed-off-by: Marco R. Pérez <marco.perez@mobica.com>

### CODE REVIEW

File: sdk/tests/deqp/functional/gles3/es3fBufferCopyTests.js [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+        ref.setSize(size);
+
+        // Create buffer.
+        var buf = this.genBuffer();
+        gl.bindBuffer(this.m_srcTarget, buf);
+
+        /** @type {Array<{srcOffset: number, dstOffset: number, copySize: number}>} */
+        var copyRanges = [{
+                srcOffset: 57, dstOffset: 701, copySize: 101 // Non-adjecent, from low to high.
```
Comment (kenrussell) [inline sdk/tests/deqp/functional/gles3/es3fBufferCopyTests.js]: typo: adjecent -> adjacent

File: sdk/tests/deqp/functional/gles3/es3fBufferCopyTests.js [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+        this.addChild(basicGroup);
+
+        for (var srcTargetNdx = 0; srcTargetNdx < bufferTargets.length; srcTargetNdx++) {
+            for (var dstTargetNdx = 0; dstTargetNdx < bufferTargets.length; dstTargetNdx++) {
+                if (srcTargetNdx == dstTargetNdx)
+                    continue;
+
+                // In WebGL 2, a copy between an ELEMENT_ARRAY_BUFFER and other data buffer
+                // (not COPY_WRITE_BUFFER nor COPY_READ_BUFFER nor ELEMENT_ARRAY_BUFFER)
+                // cannot be made, so let's skip those cases.
```
Comment (kenrussell) [inline sdk/tests/deqp/functional/gles3/es3fBufferCopyTests.js]: Great.

File: sdk/tests/deqp/functional/gles3/es3fBufferCopyTests.js [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+        for (var srcTargetNdx = 0; srcTargetNdx < bufferTargets.length; srcTargetNdx++) {
+            for (var dstTargetNdx = 0; dstTargetNdx < bufferTargets.length; dstTargetNdx++) {
+                if (srcTargetNdx == dstTargetNdx)
+                    continue;
+
+                // In WebGL 2, we can't rebind an ELEMENT_ARRAY_BUFFER or TRANSFORM_FEEDBACK_BUFFER as a
+                // different type of buffer, so we skip those cases.
+                if (bufferTargets[srcTargetNdx] == gl.ELEMENT_ARRAY_BUFFER || bufferTargets[srcTargetNdx] == gl.TRANSFORM_FEEDBACK_BUFFER ||
+                    bufferTargets[dstTargetNdx] == gl.ELEMENT_ARRAY_BUFFER || bufferTargets[dstTargetNdx] == gl.TRANSFORM_FEEDBACK_BUFFER)
```
Comment (kenrussell) [inline sdk/tests/deqp/functional/gles3/es3fBufferCopyTests.js]: Thanks for proactively modifying these tests to obey the WebGL 2.0 restrictions.

File: sdk/tests/deqp/modules/shared/glsBufferTestUtil.js [hunk 4]
```diff
@@ -0,0 +1,1068 @@
...
+     */
+    glsBufferTestUtil.getBufferTargetName = function(target) {
+        switch (target) {
+            case gl.ARRAY_BUFFER: return 'array';
+            case gl.COPY_READ_BUFFER: return 'copy_read';
+            case gl.COPY_WRITE_BUFFER: return 'copy_write';
+            case gl.ELEMENT_ARRAY_BUFFER: return 'element_array';
+            case gl.PIXEL_PACK_BUFFER: return 'pixel_pack';
+            case gl.PIXEL_UNPACK_BUFFER: return 'pixel_unpack';
+            //case gl.TEXTURE_BUFFER: return "texture"; //TODO: Unimplemented in WebGL 2. Remove?
```
Comment (kenrussell) [inline sdk/tests/deqp/modules/shared/glsBufferTestUtil.js]: This target isn't in ES 3.0, either.

File: sdk/tests/deqp/modules/shared/glsBufferTestUtil.js [hunk 5]
```diff
@@ -0,0 +1,1068 @@
...
+    /**
+     * @constructor
+     * @param {glsBufferTestUtil.WriteType} writeType
+     */
+    glsBufferTestUtil.BufferWriter = function(writeType) {
+        /** @type {glsBufferTestUtil.BufferWriterBase} */ this.m_writer = null;
+        switch (writeType) {
+            case glsBufferTestUtil.WriteType.BUFFER_SUB_DATA: this.m_writer = new glsBufferTestUtil.BufferSubDataWriter(); break;
+            default:
+                testFailed('Unsupported writer');
```
Comment (kenrussell) [inline sdk/tests/deqp/modules/shared/glsBufferTestUtil.js]: Sounds good -- I doubt that WRITE_MAP will be supported in WebGL 2.0.

Will the other ways of writing buffers (transform feedback, pixel pack) be used in subsequent tests?

File: sdk/tests/deqp/modules/shared/glsBufferTestUtil.js [hunk 6]
```diff
@@ -0,0 +1,1068 @@
...
+     * @param {WebGLBuffer} buffer
+     * @param {Uint8Array} reference
+     * @param {number} offset
+     * @param {number} numBytes
+     * @param {number} targetHint
+     * @return {boolean}
+     */
+    glsBufferTestUtil.BufferVerifierBase.prototype.verify = function(buffer, reference, offset, numBytes, targetHint) {
+        //In WebGL 2, ELEMENT_ARRAY_BUFFER and TRANSFORM_FEEDBACK_BUFFER cannot be rebound to a different
+        //type of buffer, so, let's copy their data to an ARRAY BUFFER and pass that one instead to be verified.
```
Comment (kenrussell) [inline sdk/tests/deqp/modules/shared/glsBufferTestUtil.js]: Actually the restriction is only on ELEMENT_ARRAY_BUFFERs -- buffer objects initially bound to that binding point can't be bound to other binding points. See https://www.khronos.org/registry/webgl/specs/latest/2.0/#BUFFER_OBJECT_BINDING . If you're finding other behavior (in particular, for TRANSFORM_FEEDBACK_BUFFER) that's probably a bug.

**Comment by kenrussell** [comment]
Thank you. Looks excellent overall. A few minor comments above which can be addressed in follow-on pull requests. Merging now so we can start running this test.

**Comment by marcorperez** [comment]
Thanks Ken, we will address the typos and things you mention. However,
regarding the rebinding of the TRANSFORM_FEEDBACK_BUFFER, I read this part,
which I am not too sure about, but that's why I also excluded that buffer
type from the tests, which were failing the same as with the
ELEMENT_ARRAY_BUFFER (INVALID_OPERATION):

(from https://www.khronos.org/registry/webgl/specs/latest/2.0/#5.1)
"In addition, a buffer can not be simultaneously bound to a transform
feedback buffer binding point and another buffer binding point in the WebGL
2 API. Attempting to violate this rule generates an INVALID_OPERATION
error, and the state of the binding point will remain untouched."

Marco Pérez
Mobica Ltd
Sr. Software Engineer
Mob: +52 3317 897600 <%2B44%20844%20875%201848>
Skype: mpez.mobica
www.mobica.com
<http://www.google.com/url?q=http%3A%2F%2Fwww.mobica.com%2F&sa=D&sntz=1&usg=AFrqEzeV0bYtmjTBtTc7DNhewmcdbKR9VA>

Mobica is a provider of innovative, cutting-edge software engineering,
testing and consultancy services. Based in the UK, Poland, the USA and
Mexico, Mobica has a worldwide customer base and a proven track record in
delivering complex solutions to global leaders in a range of sectors
including automotive, mobile, semiconductor, finance, TV & broadcasting,
telecommunications, connected devices, marine and aviation.

**Comment by marcorperez** [comment]
+Janusz

Marco Pérez
Mobica Ltd
Sr. Software Engineer
Mob: +52 3317 897600 <%2B44%20844%20875%201848>
Skype: mpez.mobica
www.mobica.com
<http://www.google.com/url?q=http%3A%2F%2Fwww.mobica.com%2F&sa=D&sntz=1&usg=AFrqEzeV0bYtmjTBtTc7DNhewmcdbKR9VA>

Mobica is a provider of innovative, cutting-edge software engineering,
testing and consultancy services. Based in the UK, Poland, the USA and
Mexico, Mobica has a worldwide customer base and a proven track record in
delivering complex solutions to global leaders in a range of sectors
including automotive, mobile, semiconductor, finance, TV & broadcasting,
telecommunications, connected devices, marine and aviation.

**Comment by zhenyao** [comment]
The transform feedback restraint is added to avoid undefined values.  I just added a rationale note to the spec under that restraint to make it clear.

**Comment by marcorperez** [comment]
Alright, thanks, zhenyao.

So, if i understand correctly, that's different from the ES specification
and in this case it's ok to have left that case out, right?

Marco Pérez
Mobica Ltd
Sr. Software Engineer
Mob: +52 3317 897600 <%2B44%20844%20875%201848>
Skype: mpez.mobica
www.mobica.com
<http://www.google.com/url?q=http%3A%2F%2Fwww.mobica.com%2F&sa=D&sntz=1&usg=AFrqEzeV0bYtmjTBtTc7DNhewmcdbKR9VA>

Mobica is a provider of innovative, cutting-edge software engineering,
testing and consultancy services. Based in the UK, Poland, the USA and
Mexico, Mobica has a worldwide customer base and a proven track record in
delivering complex solutions to global leaders in a range of sectors
including automotive, mobile, semiconductor, finance, TV & broadcasting,
telecommunications, connected devices, marine and aviation.

**Comment by zhenyao** [comment]
Correct.

**Pull request merged**